key.go

```diff
@@ -0,0 +1,778 @@
+
+	// NIST P-256 also known as secp256r1
+	CurveP256 Curve = 1
+
+	// NIST P-384 also known as secp384r1
+	CurveP384 Curve = 2
+
+	// NIST P-521 also known as secp521r1
+	CurveP521 Curve = 3
+
+	// X25519 for use w/ ECDH only
+	CurveX25519 Curve = 4
+
+	// X448 for use w/ ECDH only
+	CurveX448 Curve = 5
+
+	// Ed25519 for use /w EdDSA only
+	CurveEd25519 Curve = 6
+
+	// Ed448 for use /w EdDSA only
+	CurveEd448 Curve = 7
+)
+
+// Curve reprsents the EC2/OKP key's curve. See:
+// https://datatracker.ietf.org/doc/html/rfc8152#section-13.1
+type Curve int64
+
+func CurveFromString(v string) (Curve, error) {
+	switch v {
+	case "P-256":
+		return CurveP256, nil
+	case "P-384":
+		return CurveP384, nil
+	case "P-521":
+		return CurveP521, nil
+	case "X25519":
+		return CurveX25519, nil
+	case "X448":
+		return CurveX448, nil
+	case "Ed25519":
+		return CurveEd25519, nil
+	case "Ed448":
+		return CurveEd448, nil
+	default:
+		return CurveInvalid, fmt.Errorf("unknown curve value %q", v)
+	}
+}
+
+func (c Curve) String() string {
+	switch c {
+	case CurveP256:
+		return "P-256"
+	case CurveP384:
+		return "P-384"
+	case CurveP521:
+		return "P-521"
+	case CurveX25519:
+		return "X25519"
+	case CurveX448:
+		return "X448"
+	case CurveEd25519:
+		return "Ed25519"
+	case CurveEd448:
+		return "Ed448"
+	default:
+		return "unknown curve value " + strconv.Itoa(int(c))
+	}
+}
+
+// MarshalCBOR marshals the KeyType as a CBOR int.
+func (c Curve) MarshalCBOR() ([]byte, error) {
+	return encMode.Marshal(int(c))
+}
+
+// UnmarshalCBOR populates the KeyType form the provided CBOR value (must be
+// int or tstr).
+func (c *Curve) UnmarshalCBOR(data []byte) error {
+	var raw IntOrStr
+
+	if err := raw.UnmarshalCBOR(data); err != nil {
+		return fmt.Errorf("invalid curve value: %w", err)
+	}
+
+	if raw.IsString() {
+		v, err := CurveFromString(raw.String())
+
+		if err != nil {
+			return err
+		}
+
+		*c = v
+	} else {
+		v := raw.Int()
+
+		if v < 1 || v > 7 {
+			return fmt.Errorf("unknown curve value %d", v)
+		}
+
+		*c = Curve(v)
+	}
+
+	return nil
+}
+
+// Key represents a COSE_Key structure, as defined by RFC8152.
+// Note: currently, this does NOT support RFC8230 (RSA algorithms).
+type Key struct {
+	keyStruct
+	// Curve is EC identifier -- taken form "COSE Elliptic Curves" IANA registry.
+	// Populated from keyStruct.RawKeyParam when key type is EC2 or OKP.
+	Curve Curve
+	// K is the key value. Populated from keyStruct.RawKeyParam when key
+	// type is Symmetric.
+	K []byte
+}
+
+// keyStruct embedded inside Key to enable two-tier unmarshalling, of KeyType
+// and parameter with label -1 (Curve or K, depending on key type).
+type keyStruct struct {
+	// Common parameters. These are independent of the key type. Only
+	// KeyType common parameter MUST be set.
+
+	// KeyType identifies the family of keys for this structure, and thus,
+	// which of the key-type-specific parameters need to be set.
+	KeyType KeyType `cbor:"1,keyasint" json:"kty"`
```

Comment:
Still not convinced. Why do you want to support JSON encoding for this type? We don't support that for other types, and nobody complained. 

IMO, if CBOR doesn't define a secure way to encode a COSE_Key using JSON, then neither should we. Else we would be extending the spec unilaterally and without really thinking about the security implications that it could have.

Reply:
> The spec does not explicitly say anything about JSON encoding, however, these names _are_ defined by the spec, see: https://datatracker.ietf.org/doc/html/rfc8152#section-7.1

True, but they are just names (i.e., somewhat meaningful identifiers) and although there is a clear JWK pedigree here - and maybe even a desire for full compatibility in the mind of the designers - I agree with Quim that JWKs are outside COSE scope.

I guess it'd be perfectly fine if we provided a "crypto key formats" package.